cmd/soroban-cli/src/commands/contract/info/shared.rs

```diff
@@ -84,18 +113,25 @@ pub async fn fetch_wasm(args: &Args) -> Result<Option<Vec<u8>>, Error> {
             .verify_network_passphrase(Some(&network.network_passphrase))
             .await?;
 
-        get_remote_wasm_from_hash(&client, &hash).await?
+        print.globeln(format!(
+            "Downloading contract spec for wasm hash: {wasm_hash}"
+        ));
+        Ok((
+            Some(get_remote_wasm_from_hash(&client, &hash).await?),
+            None,
+            Some(network.clone()),
+        ))
     } else if let Some(contract_id) = &args.contract_id {
         let contract_id =
             contract_id.resolve_contract_id(&args.locator, &network.network_passphrase)?;
+        let contract_address = xdr::ScAddress::Contract(xdr::Hash(contract_id.0)).to_string();
+        print.globeln(format!("Downloading contract spec: {contract_address}"));
         let res = wasm::fetch_from_contract(&contract_id, network).await;
         if let Some(ContractIsStellarAsset) = res.as_ref().err() {
-            return Ok(None);
+            return Ok((None, Some(contract_address), Some(network.clone())));
```

Comment:
This return value of the function, both today and before this PR is quite confusing. Before this PR, the function is confusing because it uses an `Option` to describe a union in the `Ok`. It requires a reader to dig into the code and understand it fully to understand the intent of the meaning of the return values
- `Ok(None)` – A SAC contract
- `Ok(Some(...))` – A wasm contract
- `Err(...)` – something went wrong

This PR increases the number of things a reader must understand by reading the code:
- `Ok((Some(...), None, None))` – A wasm contract loaded from file
- `Ok((Some(...), None, Some(...))` – A wasm contract loaded from network, not looked up via a contract ID
- `Ok((Some(...), Some(...), Some(...))) – A wasm contract loaded from network, looked up via a contract ID
- `Err(...)` – something went wrong

---

I don't think this needs fixing in this PR, since it seems things were already unclear, but I think we should follow up with a change where the function is renamed from `fetch_wasm` to `fetch`, and the return is changed to `Result<Fetched, ...>`:
```rust
pub fn fetch(args: &Args, print: &Print) -> Result<Fetched, Error> { ... }

pub struct Fetched {
    contract: Contract,
    source: Source,
}

pub enum Contract {
    Wasm { wasm_bytes: Vec<u8> },
    StellarAssetContract,
}

pub enum Source {
    File { path: PathBuf },
    Wasm { hash: ..., network: Network },
    Contract { contract_id: String, network: Network },
}

impl Source {
    pub fn network(&self) -> Option<&Network> {
        match self {
            FetchedFrom::File => None,
            FetchedFrom::WasmHash { ref network, .. } => Some(network),
            FetchedFrom::Contract { ref network, .. } => Some(network),
        }
    }
}
```